grovel/grovel.lisp

```diff
@@ -179,68 +195,140 @@ int main(int argc, char**argv) {
+  (let ((fs-cache-dir (feature-specific-cache-dir
+                       cache-dir feature-expressions)))
+    (when fs-cache-dir
+      (subpathname fs-cache-dir (pathname-name file-name)
+                   :type (pathname-type file-name)))))
+
+(defun ensure-fresh-dir (abs-path)
+  (assert (uiop:absolute-pathname-p abs-path))
+  (assert (uiop:directory-pathname-p abs-path))
+  (when (uiop:directory-exists-p abs-path)
+    (uiop:delete-directory-tree abs-path :validate t))
+  (ensure-directories-exist abs-path))
+
+(defun copy-local-includes-to-cache (in-dir)
+  (assert (boundp '*local-includes*))
+  (assert *local-includes*)
+  (let ((local-include-dir
+         (uiop:ensure-directory-pathname
+          (subpathname (uiop:pathname-directory-pathname in-dir)
+                       "local-includes"))))
+    (ensure-fresh-dir local-include-dir)
+    (loop :for (src-file . local-name) :in *local-includes*
+       :for dest-file := (subpathname local-include-dir local-name)
+       :do (alexandria:copy-file src-file dest-file))
+    (format nil "-I~A" (truename local-include-dir))))
+
+
+;; This function is here to maintain backwards compatibility and is used by
+;; tests.
+(defun process-grovel-file (input-file &optional (output-defaults input-file))
+  (let* ((c-file (make-c-file-name output-defaults "__grovel"))
+         (exe-file (make-exe-file-name c-file))
+         (lisp-file (tmp-lisp-file-name c-file)))
+    (process-grovel-file* input-file lisp-file c-file exe-file nil)))
 
 ;;; *PACKAGE* is rebound so that the IN-PACKAGE form can set it during
 ;;; *the extent of a given grovel file.
-(defun process-grovel-file (input-file &optional (output-defaults input-file))
+(defun process-grovel-file* (input-file dest-lisp-file c-file exe-file absolute-cache-dir)
   (with-standard-io-syntax
-    (let* ((c-file (generate-c-file input-file output-defaults))
-           (exe-file (make-exe-file-name c-file))
-           (lisp-file (tmp-lisp-file-name c-file))
-           (inputs (list (cc-include-grovel-argument) c-file)))
-      (handler-case
-          (link-executable exe-file inputs)
-        (error (e)
-          (grovel-error "~a" e)))
-      (invoke exe-file lisp-file)
-      lisp-file)))
+    (multiple-value-bind (forms feature-expressions)
+        (read-grovel-file input-file)
+      (let* ((cached-lisp-file (feature-specific-cache-file
+                                dest-lisp-file absolute-cache-dir
+                                feature-expressions))
+             (*local-includes* nil))
+        (if (and cached-lisp-file (file-exists-p cached-lisp-file))
+            (process-grovel-file-from-cache dest-lisp-file c-file
+                                            cached-lisp-file)
+            (process-grovel-file-from-scratch forms dest-lisp-file c-file
+                                              exe-file cached-lisp-file))
+        dest-lisp-file))))
```

Comment:
We now have two cases, either:
- A `cache-dir` has been specified and both the files we need are inside it
OR
- The file or the directory are missing and we need to build them.

The first case we call `from-cache` and the second `from-scratch`